### PR TITLE
feat(yamux): future-proof public API

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

With this set of changes, we prepare the public API of `libp2p-yamux` to be as minimal as possible and allow for upgrades of the underlying `yamux` library in patch releases.

Related: #3013.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: muxers/yamux/CHANGELOG.md [hunk 1]
```diff
@@ -6,8 +6,12 @@
 - Remove deprecated items.
   See [PR 3897].
 
+- Remove `Incoming`, `LocalIncoming` and `LocalConfig` as well as anything from the underlying `yamux` crate from the public API.
+  See [PR XXXX].
+
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
 [PR 3897]: https://github.com/libp2p/rust-libp2p/pull/3897
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline muxers/yamux/CHANGELOG.md]:
```suggestion
- Remove `Incoming`, `LocalIncoming` and `LocalConfig` as well as anything from the underlying `yamux` crate from the public API.
  See [PR 3908].

[PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
[PR 3897]: https://github.com/libp2p/rust-libp2p/pull/3897
[PR 3908]: https://github.com/libp2p/rust-libp2p/pull/3908
```

File: muxers/yamux/src/lib.rs [hunk 2]
```diff
@@ -41,23 +37,25 @@ use thiserror::Error;
 use yamux::ConnectionError;
 
 /// A Yamux connection.
-pub struct Muxer<S> {
+pub struct Muxer<C> {
     /// The [`futures::stream::Stream`] of incoming substreams.
-    incoming: S,
+    incoming: BoxStream<'static, Result<yamux::Stream, yamux::ConnectionError>>,
```
Comment (mxinden) [inline muxers/yamux/src/lib.rs:42]: Why `Box` the stream of streams? Not boxing would also resolve the need for `PhantomData`, right?

**Review by mxinden — COMMENTED** [review]
Cool. Looks good to me overall. Just one suggestion.

File: muxers/yamux/src/lib.rs (hunk 2 above)
Comment (thomaseizinger) [reply muxers/yamux/src/lib.rs:42]: I just inlined what we had before.

This and the phantom data will go away once update to latest yamux. I only added it to make sure it is a compatible API change later :)

`into_stream` returns an `impl Trait`. I guess we could inline the unfold operation and store an `Unfold` type?

Do you feel strongly?

Comment (thomaseizinger) [reply muxers/yamux/src/lib.rs:42]: I am not sure the `Unfold` type is nameable at all because it contains an async block.

Comment (mxinden) [reply muxers/yamux/src/lib.rs:42]: > This and the phantom data will go away once update to latest yamux.

Great. Don't feel strongly. Let's keep as is.

**Review by mxinden — APPROVED** [review]

**Pull request merged**